torchtitan/utils.py

```diff
@@ -34,14 +34,20 @@ def get_device_info():
 device_type, device_module = get_device_info()
 
 
-def dist_max(x: Union[int, float], mesh: DeviceMesh) -> float:
-    tensor = torch.tensor(x).to(device_type)
-    return funcol.all_reduce(tensor, reduceOp=c10d.ReduceOp.MAX.name, group=mesh).item()
+def dist_reduce(x: torch.Tensor, reduceOp: str, mesh: DeviceMesh) -> float:
+    if isinstance(x, DTensor):
+        # DTensors do not support functional collectives
+        x = x.full_tensor()
```

Comment:
i am surprised by "DTensors do not support functional collectives".  I thought DTensor used functional collectives under the hood already, as part of their effort to support compiler tracing.

But in any case, doing a .full_tensor() and then using funcol is probably a worse way than calling some dtensor method that triggers dtensor to do a reduction, to save on one collective?

cc @XilunWu @yifuwang

Reply:
hm, also, what is the deal with a DTensor that has one item?  (see assertion next line) - this feels like a pretty weird case to me

Reply:
I think @carmocca meant that you can't pass a DTensor to functional collective op? I don't think we've investigated the semantic yet. Obviously it doesn't make too much sense to just perform the collective on either the local tensor or the full tensor (e.g., performing an all-gather on a replicated tensor). Something we can think about.

Reply:
> hm, also, what is the deal with a DTensor that has one item? (see assertion next line) - this feels like a pretty weird case to me

There was a request for it in https://github.com/pytorch/torchtitan/pull/779#discussion_r1909871788. Maybe renaming the functions to `scalar_reduce`, `scalar_max` ... would make the intent clearer?